salt/orch/etcd-migrate.sls

```diff
@@ -0,0 +1,155 @@
+
+update_grains:
+  salt.function:
+    - tgt: '*'
+    - name: saltutil.refresh_grains
+
+update_mine:
+  salt.function:
+    - tgt: '*'
+    - name: mine.update
+    - require:
+       - salt: update_pillar
+       - salt: update_grains
+
+update_modules:
+  salt.function:
+    - name: saltutil.sync_modules
+    - tgt: '*'
+    - kwarg:
+        refresh: True
+    - require:
+      - salt: update_mine
+
+{%- set masters = salt.saltutil.runner('mine.get', tgt='G@roles:kube-master', fun='network.ip_addrs', tgt_type='compound') %}
+{%- for master_id in masters.keys() %}
+
+{{ master_id }}-stop-services:
+  salt.state:
+    - tgt: {{ master_id }}
+    - sls:
+      - kube-apiserver.stop
+      - kube-controller-manager.stop
+      - kube-scheduler.stop
+      - docker.stop
+      - flannel.stop
+      - etcd.stop
+
+{{ master_id }}-wait-action:
+  salt.function:
+    - tgt: {{ master_id }}
+    - name: cmd.run
+    - arg:
+      - sleep 1
+    - kwarg:
+        bg: True
+    - require:
+      - salt: {{ master_id }}-stop-services
+
+{% endfor %}
+
+{%- set workers = salt.saltutil.runner('mine.get', tgt='G@roles:kube-minion', fun='network.ip_addrs', tgt_type='compound') %}
+{%- for worker_id, ip in workers.items() %}
+
+{{ worker_id }}-stop-services:
+  salt.state:
+    - tgt: {{ worker_id }}
+    - sls:
+      - kubelet.stop
+      - kube-proxy.stop
+      - docker.stop
+      - flannel.stop
+      - etcd-proxy.stop
+
+{{ worker_id }}-wait-action:
+  salt.function:
+    - tgt: {{ worker_id }}
+    - name: cmd.run
+    - arg:
+      - sleep 1
+    - kwarg:
+        bg: True
+    - require:
+      - salt: {{ worker_id }}-stop-services
+
+{% endfor %}
+
+{%- for master_id in masters.keys() %}
+
+{{ master_id }}-backup-etcd:
+  salt.function:
+    - tgt: {{ master_id }}
+    - name: cmd.run
+    - arg:
+      - mkdir /tmp/backup; btrfs subvolume snapshot /var/lib/etcd /tmp/backup/etcd
+
+{{ master_id }}-migrate-etcd:
+  salt.function:
+    - tgt: {{ master_id }}
+    - name: cmd.run
+    - arg:
+      - set -a; source /etc/sysconfig/etcdctl; env ETCD_API=3 etcdctl migrate --data-dir=/var/lib/etcd; echo "ETCD_API=3" > /var/lib/etcd/version
+
+    - require:
+      - salt: {{ master_id }}-backup-etcd
+    - kwarg:
+        bg: True
+
+{{ master_id }}-wait-etcd:
+  salt.function:
+    - tgt: {{ master_id }}
+    - name: cmd.run
+    - arg:
+      - sleep 1
+    - kwarg:
+        bg: True
+    - require:
+      - salt: {{ master_id }}-migrate-etcd
+
+{% endfor %}
+
+{%- for worker_id, ip in workers.items() %}
+
+{{ worker_id }}-backup-etcd:
+  salt.function:
+    - tgt: {{ worker_id }}
+    - name: cmd.run
+    - arg:
+      - mkdir /tmp/backup; btrfs subvolume snapshot /var/lib/etcd /tmp/backup/etcd
+
+{{ worker_id }}-migrate-etcd:
+  salt.function:
+    - tgt: {{ worker_id }}
+    - name: cmd.run
+    - arg:
+      - set -a; source /etc/sysconfig/etcdctl; env ETCD_API=3 etcdctl migrate --data-dir=/var/lib/etcd; echo "ETCD_API=3" > /var/lib/etcd/version
```

Comment:
Will this work on workers which aren't a full etcd member?

Reply:
Yes, this will work even on proxy etcd nodes

Reply:
I fixed it too